pypeit/scattlight.py

```diff
@@ -0,0 +1,169 @@
+                 'nspec': dict(otype=int, descr='Number of pixels in the image spectral direction.'),
+                 'nspat': dict(otype=int, descr='Number of pixels in the image spatial direction.'),
+                 'binning': dict(otype=str, descr='Binning in PypeIt orientation (not the original)'),
+                 'pad': dict(otype=int, descr='Integer number of pixels to mask beyond the slit edges.'),
+                 'scattlight_raw': dict(otype=np.ndarray, atype=np.floating, descr='Processed, combined scattered light image'),
+                 'scattlight_model': dict(otype=np.ndarray, atype=np.floating, descr='Model of the scattered light in scattlight_raw'),
+                 'scattlight_param': dict(otype=np.ndarray, atype=np.floating, descr='Model parameters that define the scattered light model')}
+    """Provides the class data model."""
+
+    # TODO: Allow tweaked edges to be arguments?
+    # TODO: May want nspat to be a required argument.
+    # The INIT must contain every datamodel item or risk fail on I/O when it is a nested container
+    def __init__(self, pypeline=None, detname=None, nspec=None, nspat=None, PYP_SPEC=None, binning=None, pad=0,
+                 scattlight_raw=None, scattlight_model=None, scattlight_param=None):
+
+        # Instantiate the DataContainer
+        args, _, _, values = inspect.getargvalues(inspect.currentframe())
+        _d = dict([(k,values[k]) for k in args[1:]])
+        # The dictionary passed to DataContainer.__init__ does not
+        # contain self.
+        # TODO: Does it matter if the calling function passes the
+        # keyword arguments in a different order? No.
+        datamodel.DataContainer.__init__(self, d=_d)
+
+    def _validate(self):
+        """
+        Validate the slit traces.
+        """
+        # Allow the object to be empty
+        if self.scattlight_param is None:
+            return
+
+        if self.PYP_SPEC is None:
+            self.PYP_SPEC = 'unknown'
+
+    def get_model(self, image):
+        """Generate a model of the scattered light, based on an input image. This routine
+        requires that the scattered light has already been predefined.
+
+        Parameters
+        ----------
+        image : `numpy.ndarray`_
+            A 2D image that you want to determine the amount of scattered light
+
+        Returns
+        -------
+        model : `numpy.ndarray`_
+            A model of the expected scattered light in the input image. Shape is (nspec, nspat).
+        """
+        msgs.info("Generating a scattered light image")
+        if self.scattlight_param is None:
+            msgs.warn("No scattered light parameters are available")
+            return np.zeros_like(image)
+        # Load the spectrograph
+        spec = load_spectrograph(self.PYP_SPEC)
+        # Return the model of the scattered light
+        return spec.scattered_light_model(self.scattlight_param, image)
```

Comment:
While it doesn't use anything from `self`, it could be spectrograph dependent (e.g. the number of model parameters, or some spectrographs might require the scattered light to be flipped relative to the science data, etc.). So, the model is (likely) spectrograph dependent.